### ISSUE TITLE
Provide equivalent to glslc's -M flag (output build dependencies & exit)

### ISSUE THREAD

**Issue by cdwfs** [microsoft/DirectXShaderCompiler#2063]
The glslc compiler supports a -M flag that logs dependency information for the provided source file and then exits (without attempting to compile the input). Sample output looks something like this:
```
shader.frag.spv: shader.frag localheader.h ../path/to/relativeheader.h c:\path\to/absoluteheader.h
```
I'm not married to this specific output format -- it's meant to be piped directly into a Makefile rule, which isn't a hard requirement for me.

The closest equivalent I can find for dxc is the -Vi flag, which prepends header file dependencies to the usual compilation output. Ideally I'd like to separate dependency generation from actual compilation, so I can run the former as a prepass and see which files should trigger a downstream shader to be recompiled.

**Comment by dneto0** [comment]
I believe many tools rely on the fact that the output is in Makefile syntax.
For example I think the CMake and Ninja combination use this to get precise dependency info.

So I'd try really hard to make the output format correct.  (And it's cool to see that people liked that feature of glslc. Yah!)

**Comment by ehsannas** [comment]
@cdwfs 
If we have `main.hlsl` that does `#include "a.hlsl"` and `#include "b.hlsl"`, and we are compiling `main.hlsl`, are you looking for an output like this:

```
main.hlsl: a.hlsl b:hlsl
```
or this:
```
output.spv: main.hlsl a.hlsl b.hlsl
```
?

**Comment by ehsannas** [comment]
Currently, using the `-Vi` command line option in DXC generates an output like this:

```
C:\> dxc.exe -T ps_6_0 main.hlsl -Fo result.dxil -Vi

Opening file [C:\test\a.hlsl], stack top [1]
Opening file [C:\test\b.hlsl], stack top [2]
Opening file [C:\test/c.hlsl], stack top [3]
Opening file [C:\test\d.hlsl], stack top [4]
```

@tex3d Would it be ok to change the behavior of this flag (`-Vi`) to print the dependencies in a Makefile format?

This would change the output to something like this:
```
C:\> dxc.exe -T ps_6_0 main.hlsl -Fo result.dxil -Vi

result.dxil: C:/test/a.hlsl C:/test/b.hlsl C:/test/c.hlsl C:/test/d.hlsl
```

I have a prototype (PR #2095), which only changes the behavior of `-Vi` for SPIR-V code generation cases (when `-spirv` is used). 
e.g.
```
C:\> dxc.exe -T ps_6_0 main.hlsl -Fo result.dxil -Vi

Opening file [C:\test\a.hlsl], stack top [1]
Opening file [C:\test\b.hlsl], stack top [2]
Opening file [C:\test/c.hlsl], stack top [3]
Opening file [C:\test\d.hlsl], stack top [4]

C:\> dxc.exe -T ps_6_0 main.hlsl -Fo result.spv -Vi -spirv

result.spv: C:/test/a.hlsl C:/test/b.hlsl C:/test/c.hlsl C:/test/d.hlsl
```

But it's probably better to have a consistent output regardless of the target backend. Let me know your thoughts. Thanks!

**Comment by cdwfs** [comment]
Sorry for the delay; somehow this notification eluded me for over a month! Thanks for your work on this issue.

Makefile syntax is:
`outputfile: inputfileA inputfileB ...`
The output file is the compiled .spv, not the .hlsl source file that includes the input files. So, the second option from https://github.com/Microsoft/DirectXShaderCompiler/issues/2063#issuecomment-477329270 would be correct.

`result.spv: C:/test/a.hlsl C:/test/b.hlsl C:/test/c.hlsl C:/test/d.hlsl` is close, but for completeness I believe it should also list the primary source file `main.hlsl` as a dependency.

**Comment by Wunkolo** [comment]
Any reason why this was abandoned? This was well on its way to being a valuable feature but suddenly the PR was closed suddenly.
https://github.com/microsoft/DirectXShaderCompiler/pull/2095

**Comment by dneto0** [comment]
> Any reason why this was abandoned? This was well on its way to being a valuable feature but suddenly the PR was closed. #2095 This would be very valuable in my built system to detect when edits to my "Utils.hlsl" should cause a recompilation of other shaders that might depend on it.

Good question.  Project priorities changed.

Suggest @jaebaek look at this.

**Comment by jaebaek** [comment]
Fixed by #4017

**Comment by jacobgorm** [comment]
I tried this from a master build, but it seems to not work correctly when using -Fh for output. In this case the dependency target is set to be the source file instead of the destination file, which is what Ninja or similar tools would require, and what -Fo does correctly. With -Fo it works, but I need to generate header files. -Fh and -Fo in combination do not seem to work either.

**Comment by jacobgorm** [comment]
It also seems there is a difference in semantics vs gcc (and xcrun metal) and other compilers, in that -MD appears to just dump the headers to a file and exit, whereas other compilers do this as a side effect while also compiling. So for dxc I need two invocations instead of just one to get the deps and the compiled output code.
